# 01 - Data Audit
## Product Placement Optimization - Nepali Grocery Store

### Purpose
This notebook performs a complete audit of the raw transaction dataset 
obtained from a Nepali grocery retail store. The goal is to understand 
the dataset structure, confirm data completeness, and identify quality 
issues before any cleaning or analysis begins.

### Dataset
- **Source:** Real POS transaction data from a Nepali grocery retail store
- **Size:** 768,222 rows x 14 columns
- **Period:** July 2025 to May 2026 (10 months)

In [ ]:
# Importing necessary libraries
import pandas as pd

In [ ]:
#Loading raw data
df = pd.read_excel(r'D:\softwarica\Sem 6\Individual Project\product-placement-optimization\data\raw\SALES_REGISTER.csv', header=7)
print(f"Data loaded!, Structure: {len(df)} rows and {len(df.columns)} columns.")

Data loaded!, Structure: 768222 rows and 14 columns.


In [ ]:
#First look
print(f"Columns: {list(df.columns)}")
print(f"\nFirst 5 rows:")
df.head()

Columns: ['Date', 'Miti', 'Inv.No', 'Customer', 'Product Group', 'Product', 'Uom', 'Qty', 'Rate', 'B. Amount', 'DISCOUNT', 'VAT', 'ROUND OFF', 'Total Amount']

First 5 rows:


,Date,Miti,Inv.No,Customer,Product Group,Product,Uom,Qty,Rate,B. Amount,DISCOUNT,VAT,ROUND OFF,Total Amount
0,2025-07-17,01/04/2082,ABT-000020,CASH A/C,POOJA ITEMS,Til Oil 500ml,PCS,1.0,175.0,154.867256,0.0,20.132744,0.0,175.0
1,2025-07-17,01/04/2082,ABT-000029,CASH A/C,GAULE,Gaule Chukamilo,PCS,1.0,105.0,105.000000,0.0,0.000000,0.0,105.0
2,2025-07-17,01/04/2082,ABT-000029,CASH A/C,ESSENTIAL FOOD ITEM,Krishma Mixnut 100gm,PCS,1.0,80.0,70.796460,0.0,9.203540,0.0,80.0
3,2025-07-17,01/04/2082,ABT-000029,CASH A/C,ESSENTIAL FOOD ITEM,Mong Khosta,KG,4.0,185.0,740.000000,0.0,0.000000,0.0,740.0
4,2025-07-17,01/04/2082,ABT-000029,CASH A/C,ESSENTIAL FOOD ITEM,Sugar,KG,5.0,105.0,464.601769,0.0,60.398231,0.0,525.0


In [ ]:
# Key numbers exploration
print("KEY NUMBERS ")
print("-" * 50)
print(f"Total unique invoices: {df['Inv.No'].nunique():,}")
print(f"Total unique products: {df['Product'].nunique():,}")
print(f"Unique product groups: {df['Product Group'].nunique():,}")
print(f"Unique customers:      {df['Customer'].nunique():,}")
print(f"Date range:            {df['Date'].min()} to {df['Date'].max()}")

KEY NUMBERS 
--------------------------------------------------
Total unique invoices: 218,037
Total unique products: 5,681
Unique product groups: 38
Unique customers:      25
Date range:            2025-07-17 00:00:00 to 2026-05-20 00:00:00


In [ ]:
# Product groups analysis
print("ALL PRODUCT GROUPS:")
print("-" * 70)
for i, (group,count) in enumerate(df["Product Group"].value_counts().items(),1):
    pct = (count / len(df)) * 100
    print(f"{i:2d} {group:40s} | {count:10} rows | {pct:5.1f}% |")

ALL PRODUCT GROUPS:
----------------------------------------------------------------------
 1 ESSENTIAL FOOD ITEM                      |     170669 rows |  22.2% |
 2 CANNED AND PACK GOODS                    |      96889 rows |  12.6% |
 3 SOAP AND CLEANER                         |      50868 rows |   6.6% |
 4 BISCUITS                                 |      47062 rows |   6.1% |
 5 CHOCOLATE & CANDY                        |      44774 rows |   5.8% |
 6 TEA & MASALA ITEM                        |      44202 rows |   5.8% |
 7 Cooking Oil                              |      37265 rows |   4.9% |
 8 HOUSEHOLD ITEM                           |      36091 rows |   4.7% |
 9 NOODLES ITEMS                            |      31964 rows |   4.2% |
10 SOFT DRINK                               |      30832 rows |   4.0% |
11 PERSONAL CARE                            |      29669 rows |   3.9% |
12 DAIRY PRODUCT                            |      29506 rows |   3.8% |
13 Rice                          

In [ ]:
# Top 30 products
print("TOP 30 MOST SOLD PRODUCTS:")
print("-" * 70)
for i, (products,count) in enumerate(df['Product'].value_counts().head(30).items(),1):
    print(f"{i:2d}. {products:40s} | sold{count:>7,} times |")

TOP 30 MOST SOLD PRODUCTS:
----------------------------------------------------------------------
 1. Sugar                                    | sold 23,106 times |
 2. Rato Dal                                 | sold 12,857 times |
 3. SALT                                     | sold  7,756 times |
 4. Chana                                    | sold  7,706 times |
 5. Local Milk 500ml                         | sold  7,222 times |
 6. COKE FANTA SPRITE 1.5L                   | sold  6,726 times |
 7. Kalo Dal                                 | sold  6,290 times |
 8. WAI WAI  CHICKEN                         | sold  6,171 times |
 9. Khukuri FT                               | sold  5,632 times |
10. RARA                                     | sold  5,341 times |
11. LUV POUCH DAHI 500ML                     | sold  5,209 times |
12. PATANJALI SOAP                           | sold  4,764 times |
13. Makai                                    | sold  4,760 times |
14. BADAM RAMRO                

In [ ]:
# Missing values check
print("\n1. MISSING VALUES PER COLUMN:")
nulls = df.isnull().sum()
for col, n in nulls.items():
    status = "PROBLEM" if n>0 else "OK"
    print(f"{col:25s} {n:3d} missing | {status}")


1. MISSING VALUES PER COLUMN:
Date                        1 missing | PROBLEM
Miti                        1 missing | PROBLEM
Inv.No                      1 missing | PROBLEM
Customer                    0 missing | OK
Product Group               1 missing | PROBLEM
Product                     1 missing | PROBLEM
Uom                         1 missing | PROBLEM
Qty                         1 missing | PROBLEM
Rate                        1 missing | PROBLEM
B. Amount                   1 missing | PROBLEM
DISCOUNT                  158 missing | PROBLEM
VAT                       158 missing | PROBLEM
ROUND OFF                 158 missing | PROBLEM
Total Amount                1 missing | PROBLEM


In [ ]:
# Duplicates check
dupes = df.duplicated().sum()
print(f"\n2. DUPLICATE ROWS: {dupes:,}")


2. DUPLICATE ROWS: 1,040


In [ ]:
# Negative values
neg_qty = (df['Qty'] <= 0).sum()
neg_amount = (df['Total Amount'] <= 0).sum()
print(f"\n3. ROWS WITH QTY <= 0:          {neg_qty:,}")
print(f"   ROWS WITH TOTAL AMOUNT <= 0: {neg_amount:,}")


3. ROWS WITH QTY <= 0:          0
   ROWS WITH TOTAL AMOUNT <= 0: 0


In [ ]:
# NaN in product group
nan_pg = df['Product Group'].isna().sum()
print(f"\n4. ROWS WITH NO PRODUCT GROUP:  {nan_pg:,}")



4. ROWS WITH NO PRODUCT GROUP:  1


# Audit Complete

### Summary of Findings
| Issue | Count | Action Required |
|-------|-------|----------------|
| Duplicate rows | 1,040 | Remove in cleaning |
| Missing product group | 1 | Remove in cleaning |
| Blank row | 1 | Remove in cleaning |
| Missing DISCOUNT/VAT/ROUND OFF | 158 | Investigate in cleaning |
| Inconsistent category names | Multiple | Standardize in cleaning |

### Next Step
Proceed to **02_data_cleaning.ipynb** to fix all issues identified above.

In [1]:
import pandas as pd
import os

# List all CSV files in your data folder
for root, dirs, files in os.walk('data'):
    for f in files:
        filepath = os.path.join(root, f)
        print(filepath, '---', os.path.getsize(filepath), 'bytes')
        

In [2]:
# Load your raw file and show me what it looks like
df = pd.read_csv('data/raw/SALES_REGISTER_DATA.csv')
print(df.shape)
print("---")
print(df.columns.tolist())
print("---")
print(df.dtypes)
print("---")
df.head()

FileNotFoundError: [Errno 2] No such file or directory: 'data/raw/SALES_REGISTER_DATA.csv'

In [4]:
import os
os.listdir('../data/raw')

['SALES_REGISTER.csv']

In [5]:
import pandas as pd

df = pd.read_csv('../data/raw/SALES_REGISTER.csv')
print(df.shape)
print("---")
print(df.columns.tolist())
print("---")
print(df.dtypes)
print("---")
df.head()

(218030, 13)
---
['Exported   Miti/Time : 06/02/2083( 20/05/2026 ) 06:06:53 PM ', 'Unnamed: 1', 'Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4', 'Unnamed: 5', 'Fiscal Year : 82-83 (01/04/2082-32/03/2083)', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9', 'Unnamed: 10', 'Unnamed: 11', 'Unnamed: 12']
---
Exported   Miti/Time : 06/02/2083( 20/05/2026 ) 06:06:53 PM        str
Unnamed: 1                                                         str
Unnamed: 2                                                         str
Unnamed: 3                                                         str
Unnamed: 4                                                         str
Unnamed: 5                                                         str
Fiscal Year : 82-83 (01/04/2082-32/03/2083)                        str
Unnamed: 7                                                         str
Unnamed: 8                                                         str
Unnamed: 9                                                      object


C:\Users\samek\AppData\Local\Temp\ipykernel_30664\2797165295.py:3: DtypeWarning: Columns (0: Unnamed: 9) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../data/raw/SALES_REGISTER.csv')


,Exported Miti/Time : 06/02/2083( 20/05/2026 ) 06:06:53 PM,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Fiscal Year : 82-83 (01/04/2082-32/03/2083),Unnamed: 7,Unnamed: 8,Unnamed: 9,Unnamed: 10,Unnamed: 11,Unnamed: 12
0,BANIYA SHOPPING CENTER PVT.LTD.,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,LAMACHOUR-16 POKHARA,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,PAN No : 610418608 ; Phone : 061-443259,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Sales Register,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,From 01/04/2082 ( 17/07/2025) To 06/02/2083 ...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [6]:
# Let's look at rows 0 to 15 to find where actual data starts
temp = pd.read_csv('../data/raw/SALES_REGISTER.csv', header=None, nrows=15)
for i, row in temp.iterrows():
    print(f"Row {i}: {row.tolist()}")

Row 0: ['Exported   Miti/Time : 06/02/2083( 20/05/2026 ) 06:06:53 PM ', nan, nan, nan, nan, nan, 'Fiscal Year : 82-83 (01/04/2082-32/03/2083)', nan, nan, nan, nan, nan, nan]
Row 1: ['BANIYA SHOPPING CENTER  PVT.LTD.', nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]
Row 2: ['LAMACHOUR-16 POKHARA', nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]
Row 3: ['PAN No : 610418608 ; Phone : 061-443259', nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]
Row 4: ['Sales Register  ', nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]
Row 5: ['From 01/04/2082 ( 17/07/2025)  To  06/02/2083 ( 20/05/2026)', nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]
Row 6: ['Date', 'Miti', 'Inv.No', 'Customer', 'Billing Party', 'Qty', 'B. Amount', 'DISCOUNT Amt', 'VAT Amt', 'ROUND OFF', 'Total Amount', 'Cash/Credit', 'BillType']
Row 7: ['22/07/2025', '06/04/2082', 'ABT-003614', 'CASH A/C', 'CASH A/C', '2.00', '440.00', '0.00', '0.00', '0.00', '440.00', 'Cash', 

In [7]:
# Check the sample_data file
df2 = pd.read_csv('../data/sample_data.csv')
print(df2.shape)
print("---")
print(df2.columns.tolist())
print("---")
df2.head()

(42, 14)
---
['Date', 'Miti', 'Inv.No', 'Customer', 'Product Group', 'Product', 'Uom', 'Qty', 'Rate', 'B. Amount', 'DISCOUNT', 'VAT', 'ROUND OFF', 'Total Amount']
---


,Date,Miti,Inv.No,Customer,Product Group,Product,Uom,Qty,Rate,B. Amount,DISCOUNT,VAT,ROUND OFF,Total Amount
0,17/07/2025,1/4/2082,ABT-000020,CASH A/C,POOJA ITEMS,Til Oil 500ml,PCS,1,175,154.87,0,20.13,0,175
1,17/07/2025,1/4/2082,ABT-000029,CASH A/C,GAULE,Gaule Chukamilo,PCS,1,105,105,0,0.00,0,105
2,17/07/2025,1/4/2082,ABT-000029,CASH A/C,ESSENTIAL FOOD ITEM,Krishma Mixnut 100gm,PCS,1,80,70.8,0,9.20,0,80
3,17/07/2025,1/4/2082,ABT-000029,CASH A/C,ESSENTIAL FOOD ITEM,Mong Khosta,KG,4,185,740,0,0.00,0,740
4,17/07/2025,1/4/2082,ABT-000029,CASH A/C,ESSENTIAL FOOD ITEM,Sugar,KG,5,105,464.6,0,60.40,0,525


In [8]:
import os
for root, dirs, files in os.walk('..'):
    for f in files:
        if f.endswith('.csv'):
            print(os.path.join(root, f))

..\data\sample_data.csv
..\data\raw\SALES_REGISTER.csv
..\venv\Lib\site-packages\matplotlib\mpl-data\sample_data\data_x_x2_x3.csv
..\venv\Lib\site-packages\matplotlib\mpl-data\sample_data\msft.csv
..\venv\Lib\site-packages\matplotlib\mpl-data\sample_data\Stocks.csv
..\venv\Lib\site-packages\mlxtend\data\data\boston_housing.csv
..\venv\Lib\site-packages\mlxtend\data\data\wine.csv
..\venv\Lib\site-packages\numpy\random\tests\data\mt19937-testset-1.csv
..\venv\Lib\site-packages\numpy\random\tests\data\mt19937-testset-2.csv
..\venv\Lib\site-packages\numpy\random\tests\data\pcg64-testset-1.csv
..\venv\Lib\site-packages\numpy\random\tests\data\pcg64-testset-2.csv
..\venv\Lib\site-packages\numpy\random\tests\data\pcg64dxsm-testset-1.csv
..\venv\Lib\site-packages\numpy\random\tests\data\pcg64dxsm-testset-2.csv
..\venv\Lib\site-packages\numpy\random\tests\data\philox-testset-1.csv
..\venv\Lib\site-packages\numpy\random\tests\data\philox-testset-2.csv
..\venv\Lib\site-packages\numpy\random\tests

In [9]:
df = pd.read_csv('../data/raw/SALES_REGISTER.csv', skiprows=6)
print(df.shape)
print("---")
print(df.columns.tolist())
print("---")
print(df.dtypes)
print("---")
df.head()

(218024, 13)
---
['Date', 'Miti', 'Inv.No', 'Customer', 'Billing Party', 'Qty', 'B. Amount', 'DISCOUNT Amt', 'VAT Amt', 'ROUND OFF', 'Total Amount', 'Cash/Credit', 'BillType']
---
Date                 str
Miti                 str
Inv.No               str
Customer             str
Billing Party        str
Qty                  str
B. Amount            str
DISCOUNT Amt         str
VAT Amt              str
ROUND OFF        float64
Total Amount         str
Cash/Credit          str
BillType             str
dtype: object
---


,Date,Miti,Inv.No,Customer,Billing Party,Qty,B. Amount,DISCOUNT Amt,VAT Amt,ROUND OFF,Total Amount,Cash/Credit,BillType
0,22/07/2025,06/04/2082,ABT-003614,CASH A/C,CASH A/C,2.00,440.00,0.00,0.00,0.0,440.00,Cash,Abbreviated Tax Invoice
1,22/07/2025,06/04/2082,ABT-003623,CASH A/C,CASH A/C,6.00,865.00,0.00,74.78,0.0,865.00,Cash,Abbreviated Tax Invoice
2,22/07/2025,06/04/2082,ABT-003634,CASH A/C,CASH A/C,2.00,120.00,0.00,0.00,0.0,120.00,Cash,Abbreviated Tax Invoice
3,22/07/2025,06/04/2082,ABT-003648,CASH A/C,CASH A/C,6.50,"1,110.00",0.00,72.48,0.0,"1,110.00",Cash,Abbreviated Tax Invoice
4,22/07/2025,06/04/2082,ABT-003657,CASH A/C,CASH A/C,1.00,100.00,0.00,11.50,0.0,100.00,Cash,Abbreviated Tax Invoice
